### ISSUE TITLE
Synchronizing client ids between Clients and ClientManager

### ISSUE THREAD

**Issue by lguerdan** [adap/flower#550]
Hi there, I’m using Flower to set up a federated learning experiment and came across a question — I’m new to the framework and apologize if this is something simple that I missed. 

In an experiment I am setting up, I am configuring a `ClientManager` to perform client selection based on client-specific configuration parameters (for example, a client-specific drop probability that approximates a bad network connection). In addition to this server-side selection process, the client-specific configuration will also control client dataset properties that are loaded in the `Client` class. For this to work, I need to be able to link each `Client` `cid` on the client side with each `ClientProxy` `cid` in the server `ClientManager`. 

However, an issue that I am encountering is that the `ClientProxy` `cids` reflect `GrpcClientProxy` ids (i.e. `ipv6:[::1]:5589`) rather than the numeric ids I assign while initializing the clients (i.e. `1`). This is an issue because the `ClientManager` `register` function will assign clients as they arrive, and this order may not match the index of configurations loaded in the `Client` class. Therefore I believe a common ID mapping the `Clients` and `ClientProxies` would be necessary to achieve this setup. 

Upon closer inspection of the `start_client` in `app.py`, I don’t see any communication of the `Client` `cid` via the grpc server, so this may be the intended behavior. I also checked existing strategies (`fedfs_v0`, `fast_and_slow`, etc.), and while these strategies appear too perform selection based on client training metadata, the information is based on the server-side strategy and previous rounds, and doesn’t require linking back to the client processes. 

Do you have any recommended workarounds for supporting this type of behavior? I can think of a few workarounds achieve it, but wanted to check whether it is supported in the framework / whether I am missing something first. 

Thanks for the help! The framework has been great to use so far.

**Comment by danieljanes** [comment]
Hi @lguerdan , glad to hear you like the framework so far! What you describe is quite accurate, the server-side `cid` is not linked to any client-side property.

This is one area that we want to improve: we want the client to be able to communicate metadata to the server when it establishes the connection (e.g., "I'm client no. 5 and I have a GPU"). This would allow the server to perform sampling that takes advantage of this metadata (e.g., "select only GPU-accelerated clients"). The client should be able to communicate arbitrary metadata, so it'll be up to the framework user to decide what should be included (platform, hardware capabilities, dropout probabilities, etc.). Would this be helpful for you?

Until that's available you can do two things (warning: only viable for simulation purposes and a little bit hacky):
1. Simulate different client by "reconfiguring" them based on your `Strategy`: you implement your own sampling in a custom strategy the get, for example, ten client IDs for the next round. You then ask the ClientManager for ten random clients and when you build the `FitIns` object, you include the IDs in the `config` dictionary. When one of the sampled clients receives the `FitIns` object, it first gets the `ID` from the `config` dict and uses this ID to "reconfigure" itself (e.g., to load the appropriate data partition for the respective client ID). This essentially means that the running clients processes switch their identity to whatever client ID is currently selected for training.
2. Create some kind of mapping between server-side `cid` and some client-side identifier. There are many (hacky) ways to do this, but essentially you could start the clients in a particular order (with some time delay in between) and do the following: (a) client 0 starts and writes 0 into a file (b) server receives connection from client 0, assigns it a `cid`, and writes this `cid` to the same file - we now know which client-identifier belongs to which server-side `cid` - (c) do this every time a client reconnects to the server to keep the file updated and use this file in your strategy implementation to improve your sampling algorithm

I hope this helps. Again, both suggestions are rather hacky and not really nice to work with, which is why we're planning to build a much nicer general-purpose way of doing this :)

**Comment by lguerdan** [comment]
Thanks for the thorough response @danieljanes , I'm glad that I wasn't just over complicating things :) 

> This is one area that we want to improve: we want the client to be able to communicate metadata to the server when it establishes the connection (e.g., "I'm client no. 5 and I have a GPU"). This would allow the server to perform sampling that takes advantage of this metadata (e.g., "select only GPU-accelerated clients"). The client should be able to communicate arbitrary metadata, so it'll be up to the framework user to decide what should be included (platform, hardware capabilities, dropout probabilities, etc.). Would this be helpful for you?

Yes, this would be fantastic. I like the current setup with `FitIns` and and `EvaluateIns` where you have a `map<string, string> config` in the protocol buffer specification, then enable adding additional configs dynamically on the client. Something similar in the connection object linking the client to the server would be great and would have been perfect for what I am trying to achieve here.  

As a side note, adding this on the `Fit` / `Eval` response would also be helpful. I added a new return parameter in `FitRes` and had to fork and recompile the protocol buffer to support that. Although looks like you guys already have that planned [here](https://github.com/adap/flower/issues/489) .

> Until that's available you can do two things (warning: only viable for simulation purposes and a little bit hacky):

I hadn't thought about these approaches, I think either of them would work. I ended up coming up with something simple for the narrow drop probability case yesterday. I just pass in a drop probability as a FitIn and have the client return a failure based on the drop probability. Because the data partition and fit config are both provided in the client this works fine for this narrow workaround. I think if I wanted anything that actually links a server with the client (1.) or (2.) would also be a good temporary solution.  

Thanks for the help! Overall the framework has saved a lot of time for getting started with FL, I appreciate it.

**Comment by danieljanes** [comment]
Thanks for the update @lguerdan, glad to hear you found a good solution. Closing this issue now, feel free to re-open if something comes up!

**Comment by danieljanes** [comment]
Hi @lguerdan, our implementations for #489 and #490 were merged into the `main` branch, the features are available in the latest nightly builds (`pip install flw-nightly`). We're curious to get your thoughts on the solution, is it general enough for your use case so that you don't need to change/re-compile the `.proto` anymore?

You can find a  written description of the approach in the latest release notes ("Unreleased"): https://flower.dev/docs/changelog.html

**Comment by lguerdan** [comment]
Hey @danieljanes thanks for adding this to the new release, this is great. I haven't installed the nightly build yet, but based on the release notes this would handle the use case I have and prevent me from needing to re-compile the `.proto`. Generalizing the return types beyond strings is also a nice update.

It may be helpful to roll the `accuracy` in the `EvaluateRes` into the custom config dictionary (some of the use-cases I am considering involve regression), but ultimately populating that with a dummy value if it's unused is fine. You're probably leaving that as-is for backwards compatibility as well. 

Great to see the framework continue to improve!

**Comment by danieljanes** [comment]
Thanks for the feedback @lguerdan, happy to hear that it fits your requirements. You're spot on, we're preserving `accuracy` purely for compatibility reasons, but we'll drop in in an incompatible change before we release Flower 1.0.

**Comment by Firas-Kh-1** [comment]
Hi there @danieljanes @lguerdan, I read through this thread and the documentation of the framework I started using recently and I have a similar question regarding selecting clients. 

 In the project I am working on, the server is supposed to select the clients before starting the training based on their `client_id` but I couldn't find a way to use that client-side variable to make the selection on the server using `Criterion` or other functions. 
 The idea mentioned above about communicating metadata between the server and client upon connection "I am client no. 5 and I have a GPU" would be the ideal solution given the importance of client selection in federated ml so we could make selection using `ClientManager.clients.client_id` for example.
 
 I apologise if this is a simple issue that has already been solved and I missed. Any help would be appreciated!
 Thanks.
